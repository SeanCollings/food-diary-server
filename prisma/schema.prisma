// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
