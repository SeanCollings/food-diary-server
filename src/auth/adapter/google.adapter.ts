import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { HttpService } from '@nestjs/axios';
import { catchError, firstValueFrom } from 'rxjs';
import { AxiosError } from 'axios';
import { RecaptchaResponse } from '../types';
import { GOOGLE_RECAPTCHA_PASS_SCORE } from '../auth.constants';

@Injectable()
export class GoogleAdapter {
  private recaptchaVerifyUrl: string;
  private recaptchaSecretKey: string;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.recaptchaVerifyUrl =
      this.configService.get('RECAPTCHA_VERIFY_URL') || '';
    this.recaptchaSecretKey =
      this.configService.get('RECAPTCHA_SECRET_KEY') || '';
  }

  async verifySite(token: string) {
    const params = {
      secret: this.recaptchaSecretKey,
      response: token,
    };

    const { data } = await firstValueFrom(
      this.httpService
        .post<RecaptchaResponse>(this.recaptchaVerifyUrl, null, { params })
        .pipe(
          catchError((error: AxiosError) => {
            console.error('recaptcha error:', error.response?.data);
            throw new Error('A verification error ocurred!');
          }),
        ),
    );

    if (!data.success) {
      throw new Error(
        `Verification failed for ${data.action}: ${(
          data['error-codes'] ?? []
        ).join()}`,
      );
    }

    return data.score >= GOOGLE_RECAPTCHA_PASS_SCORE;
  }
}
