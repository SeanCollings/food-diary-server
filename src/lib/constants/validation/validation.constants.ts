export const PASSWORD_MIN_LENGTH = 8;
export const EMAIL_MIN_LENGTH = 6;
export const INPUT_SMALL_MAX_LENGTH = 25;
export const INPUT_MAX_LENGTH = 100;
export const TEXTAREA_MAX_LENGTH = 256;

export const EMAIL_REGEX = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
