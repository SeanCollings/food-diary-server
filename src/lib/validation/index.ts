import { trim } from '@/utils/string-utils';
import {
  EMAIL_MIN_LENGTH,
  EMAIL_REGEX,
  INPUT_MAX_LENGTH,
  PASSWORD_MIN_LENGTH,
} from '@/lib/constants/validation/validation.constants';

export const validateEmail = (email: string) => {
  const trimEmail = trim(email);

  if (!trimEmail?.length) {
    return false;
  } else if (!EMAIL_REGEX.test(trimEmail)) {
    return false;
  }

  return (
    trimEmail.length >= EMAIL_MIN_LENGTH && trimEmail.length <= INPUT_MAX_LENGTH
  );
};

export const validatePassword = (password: string) => {
  const trimPassword = trim(password);

  if (!trimPassword?.length) {
    return false;
  }

  return (
    trimPassword.length >= PASSWORD_MIN_LENGTH &&
    trimPassword.length <= INPUT_MAX_LENGTH
  );
};

export const validateLogin = ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => validateEmail(email) && validatePassword(password);
