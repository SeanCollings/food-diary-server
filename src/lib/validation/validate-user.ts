import { promisify } from 'util';
import { scrypt as _scrypt } from 'crypto';
import { User } from '@prisma/client';
import { SCRYPT_KEYLEN } from '@/auth/auth.constants';

const scrypt = promisify(_scrypt);

export const isValidPassword = async (user: User | null, password: string) => {
  if (!user) {
    return null;
  }

  const [salt, storedHash] = user.password.split('.');
  const hash = (await scrypt(password, salt, SCRYPT_KEYLEN)) as Buffer;

  if (storedHash === hash.toString('hex')) {
    return user;
  }

  return null;
};
