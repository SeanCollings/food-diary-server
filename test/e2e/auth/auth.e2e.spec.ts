import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { PrismaClient } from '@prisma/client';
import { PrismaService } from '@/prisma.service';
import { AppModule } from '@/app.module';
import { isValidPassword } from '@/lib/validation/validate-user';
import { GoogleAdapter } from '@/auth/adapter/google.adapter';
import { JwtService } from '@nestjs/jwt';

import { user_fixture } from '../__fixtures__';

jest.mock('@/lib/validation/validate-user');

const mockError = jest.fn();

describe('AuthController (e2e)', () => {
  let app: INestApplication;
  let prisma: DeepMockProxy<PrismaClient>;
  let mockGoogleAdapter: jest.Mocked<Partial<GoogleAdapter>>;
  let mockJwtSerive: jest.Mocked<Partial<JwtService>>;

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(mockError);
  });

  beforeEach(async () => {
    mockGoogleAdapter = {
      verifySite: jest.fn(),
    };
    mockJwtSerive = {
      sign: jest.fn(),
    };

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .overrideProvider(GoogleAdapter)
      .useValue(mockGoogleAdapter)
      .overrideProvider(JwtService)
      .useValue(mockJwtSerive)
      .compile();

    app = moduleFixture.createNestApplication();
    await app.init();
    prisma = moduleFixture.get(PrismaService);
    mockGoogleAdapter = moduleFixture.get(GoogleAdapter);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/signup (POST)', () => {
    describe('positive', () => {
      it('should signup a new user', async () => {
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);
        prisma.user.findUnique.mockResolvedValue(null);
        prisma.user.create.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .post('/auth/signup')
          .send({
            name: 'New User',
            email: 'test@email.com',
            password: 'password',
            token: 'mock_token',
          });

        expect(response.status).toEqual(201);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should fail if no body sent', async () => {
        const response = await request(app.getHttpServer()).post(
          '/auth/signup',
        );

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });

      it('should fail if user name is missing', async () => {
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);
        prisma.user.findUnique.mockResolvedValue(user_fixture);
        prisma.user.create.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .post('/auth/signup')
          .send({
            name: null,
            email: 'test@email.com',
            password: 'password',
            token: 'mock_token',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('/login (POST)', () => {
    describe('positive', () => {
      it('should login a user', async () => {
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);
        (mockJwtSerive.sign as jest.Mock).mockReturnValue('mock_new_token');
        prisma.user.update.mockResolvedValue(user_fixture);
        prisma.user.create.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .post('/auth/login')
          .send({
            email: 'test@email.com',
            password: 'password',
            token: 'mock_token',
          });

        expect(response.status).toEqual(201);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should fail if no body sent', async () => {
        const response = await request(app.getHttpServer()).post('/auth/login');

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should fail if password not valid', async () => {
        const response = await request(app.getHttpServer())
          .post('/auth/login')
          .send({
            email: 'test@email.com',
            password: '',
            token: 'mock_token',
          });

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should fail if email not valid', async () => {
        const response = await request(app.getHttpServer())
          .post('/auth/login')
          .send({
            email: '@email.com',
            password: 'password',
            token: 'mock_token',
          });

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should fail if google captcha fails', async () => {
        (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
        prisma.user.findUnique.mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(false);

        const response = await request(app.getHttpServer())
          .post('/auth/login')
          .send({
            email: 'test@email.com',
            password: 'password',
            token: 'mock_token',
          });

        expect(response.status).toEqual(500);
        expect(response.body).toMatchSnapshot();
      });

      it('should fail if cant validate user', async () => {
        (isValidPassword as jest.Mock).mockResolvedValue(null);
        prisma.user.findUnique.mockResolvedValue(user_fixture);
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const response = await request(app.getHttpServer())
          .post('/auth/login')
          .send({
            email: 'test@email.com',
            password: 'password',
            token: 'mock_token',
          });

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('/reset (POST)', () => {
    describe('positive', () => {
      it('should reset a user password', async () => {
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

        const response = await request(app.getHttpServer())
          .post('/auth/reset')
          .send({
            email: 'test@email.com',
            token: 'mock_token',
          });

        expect(response.status).toEqual(201);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should fail if google captcha fails', async () => {
        (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(false);

        const response = await request(app.getHttpServer())
          .post('/auth/reset')
          .send({
            email: 'test@email.com',
            token: 'mock_token',
          });

        expect(response.status).toEqual(500);
        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
