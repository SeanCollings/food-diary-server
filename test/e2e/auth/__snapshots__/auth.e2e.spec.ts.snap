// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController (e2e) /login (POST) negative should fail if cant validate user 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`AuthController (e2e) /login (POST) negative should fail if email not valid 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`AuthController (e2e) /login (POST) negative should fail if google captcha fails 1`] = `
{
  "error": "Internal Server Error",
  "message": "Something went wrong!",
  "statusCode": 500,
}
`;

exports[`AuthController (e2e) /login (POST) negative should fail if no body sent 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`AuthController (e2e) /login (POST) negative should fail if password not valid 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`AuthController (e2e) /login (POST) positive should login a user 1`] = `
{
  "access_token": "mock_new_token",
}
`;

exports[`AuthController (e2e) /reset (POST) negative should fail if google captcha fails 1`] = `
{
  "error": "Internal Server Error",
  "message": "Something went wrong!",
  "statusCode": 500,
}
`;

exports[`AuthController (e2e) /reset (POST) positive should reset a user password 1`] = `
{
  "message": "Reset link sent",
}
`;

exports[`AuthController (e2e) /signup (POST) negative should cater for email in use error 1`] = `
{
  "error": "Conflict",
  "message": "Email in use!",
  "statusCode": 409,
}
`;

exports[`AuthController (e2e) /signup (POST) negative should fail if no body sent 1`] = `
{
  "error": "Bad Request",
  "message": [
    "name must be shorter than or equal to 100 characters",
    "name should not be empty",
    "name must be a string",
    "email must be shorter than or equal to 100 characters",
    "email must be longer than or equal to 6 characters",
    "email should not be empty",
    "email must be an email",
    "password must be shorter than or equal to 100 characters",
    "password must be longer than or equal to 8 characters",
    "password should not be empty",
    "password must be a string",
    "token should not be empty",
    "token must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`AuthController (e2e) /signup (POST) negative should fail if user name is missing 1`] = `
{
  "error": "Bad Request",
  "message": [
    "name must be shorter than or equal to 100 characters",
    "name should not be empty",
    "name must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`AuthController (e2e) /signup (POST) positive should signup a new user 1`] = `
{
  "avatar": "",
  "email": "test@email.com",
  "name": "Mock User",
}
`;
